File: linked_list/linked_list_reverse.py
"""
linked list
"""

# *-* encoding: utf-8 -*-

import copy


class Node():
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class LinkedList():
    def __init__(self):
        self.root = None
        return

    def construct_linked_list(self):
        curr_node = self.root

        curr_node = Node(1)
        curr_node.next = Node(2)
        curr_node.next.next = Node(3)
        curr_node.next.next.next = Node(4)
        curr_node.next.next.next.next = Node(5)
        curr_node.next.next.next.next.next = Node(3)

        self.root = curr_node

    def print(self):
        curr_node = self.root
        while(curr_node is not None):
            print(curr_node.data, end=" ")
            curr_node = curr_node.next
        print()

    def reverse(self):
        prev = None
        curr = self.root
        next = curr.next

        while(next is not None):
            curr.next = prev
            prev = curr
            curr = next
            next = curr.next
        curr.next = prev

        self.root = curr
        return

    def remove_duplicates(self):
        linked_list_set = set()

        curr = self.root
        prev = None

        while(curr is not None):
            if curr.data in linked_list_set:
                prev.next = curr.next
                if prev.next is not None:
                    curr = prev.next
                else:
                    curr = None
            else:
                linked_list_set.add(curr.data)
                prev = curr
                curr = curr.next

        return


def main():
    linked_list = LinkedList()
    linked_list.construct_linked_list()
    linked_list.print()

    # Reverse linked list
    linked_list.reverse()
    linked_list.print()

    # Remove duplicates from linked list
    linked_list.remove_duplicates()
    linked_list.print()

    return

if __name__ == '__main__':
    main()

File: linked_list/linked_list_reverse.cpp
// compile using g++ -std=c++0x 2_3_4_array_sum.cpp

#include <iostream>

using namespace std;

// linkedlist Node
class Node {
    public:
        int data;
        Node* next;
        Node(int data) {
            this->data = data;
            this->next = NULL;
        }
};

class LinkedList {
    public:
        Node* head;
        LinkedList() {
            this->head = NULL;
        }

        void constructLinkedList() {
            this->head                               = new Node(1);
            this->head->next                         = new Node(2);
            this->head->next->next                   = new Node(3);
            this->head->next->next->next             = new Node(4);
            this->head->next->next->next->next       = new Node(5);
            this->head->next->next->next->next->next = new Node(6);
            return;
        }

        void print_linkedList() {
            if(this->head == NULL) {
                return;
            }

            Node* temp = this->head;
            while(temp != NULL) { 
                cout << temp->data << " ";
                temp = temp->next;
            }
            cout << endl;
            return;
        }

        // reverse linkedlist iteratively
        void iterative_reverse() {
            if((this->head == NULL) || (this->head->next == NULL)) {
                return;
            }

            Node* temp = this->head->next;
            Node* temp1 = NULL;

            while(this->head->next != NULL) {
                this->head->next = temp1;
                temp1 = this->head;
                this->head = temp;
                temp = temp->next;        
            }
            this->head->next = temp1;

            return;
        }
};

int main() {
    // make an empty linked list, and add data to it
    LinkedList linkedlist;
    linkedlist.constructLinkedList();

    // print linked list
    linkedlist.print_linkedList();

    cout << "reversing linked list" << endl;
    linkedlist.iterative_reverse();

    // print linked list
    linkedlist.print_linkedList();

    return 0;
}
